docs/proposals/disk-accounting.md

```diff
@@ -0,0 +1,641 @@
+
+	We can also attempt to use this feature to determine image layer’s disk usage statically. The metadata files that docker creates for these layers will not be tracked though.
+
+**Pros**
+
+* Works across all storage drivers since docker abstracts the storage drivers.
+
+* Less code to maintain in kubelet.
+
+**Cons**
+
+* Not available today.
+
+* Requires serialized image pulls.
+
+* Metadata files are not tracked.
+
+
+##### Devicemapper
+
+Devicemapper storage driver will setup two volumes, metadata and data, that will be used to store image layers and container writable layer. The volumes can be real devices or loopback. A Pool device is created which uses the underlying volume for real storage.
+
+A new thinly-provisioned volume, based on the pool, will be created for running container’s.
+
+The kernel tracks the usage of the pool device at the block device layer. The usage here includes image layers and container’s writable layers.
+
+Since the kubelet has to track the writable layer usage anyways, we can subtract the aggregated root filesystem usage from the overall pool device usage to get the image layer’s disk usage.
+
+Linux quota and `du` will not work with device mapper.
+
+A docker dry run option (mentioned above) is another possibility.
+
+
+#### Container Writable Layer
+
+###### Overlayfs / Aufs
+
+Docker creates a separate directory for the container’s writable layer which is then overlayed on top of read-only image layers.
+
+Both the previously mentioned options of `du` and `Linux Quota` will work for this case as well.
+
+Kubelet can use `du` to track usage and enforce `limits` once disk becomes a schedulable resource. As mentioned earlier `du` is resource intensive.
+
+To use Disk quota, kubelet will have to allocate a separate gid per container. Kubelet can reuse the same gid for multiple instances of the same container (restart scenario). As and when kubelet garbage collects dead containers, the usage of the container will drop.
+
+If local disk becomes a schedulable resource, `linux quota` can be used to impose `request` and `limits` on the container writable layer.
+`limits` can be enforced using hard limits. Enforcing `request` will be tricky. One option is to enforce `requests` only when the disk availability drops below a threshold (10%). Kubelet can at this point evict pods that are exceeding their requested space. Other options include using `soft limits` with grace periods, but this option is complex.
+
+###### Devicemapper
+
+Accessing stats requires entering the docker’s mount namespace and then fetch the filesystem stats on the fs mounted for the container’s writable layer.
+
+For example, if the container’s id is `92b94ac8c3`, then to get disk usage stats for that container, the following command needs to be run,
+
+```shell
+ nsenter -m -t 16663 stat -f /var/lib/docker/devicemapper/mnt/92b94ac8c3f3a932f6fdf30b86edcdff94d8e5f71e36b79da8419294c19eb6f2/rootfs/
+
```

Comment:
Can we afford to start the container, do statfs and then shutdown container? I think it will be racy with other tools though.

Ideally we need to have capability to just mount container rootfs and run statfs and then unmount it in such a way that it does not race with docker. There is already some facility available in "atomic" command where one can do "atomic mount". But right not that racy w.r.t docker. We are looking into improving it. So once that comes along, you can start using it.